### PR TITLE
Allow disabling JS rules in standalone and connected mode

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Part of #771

### CODE REVIEW

File: src/Core.UnitTests/SonarCompositeRuleIdTests.cs [hunk 1]
```diff
@@ -48,5 +48,15 @@ public void TryParse_Valid_ReturnExpecteed(string errorCode, string expectedRepo
...
+
+
+        [TestMethod]
+        public void ToString_ReturnsExpected()
+        {
+            const string input = "xxx:YYY";
+
+            SonarCompositeRuleId.TryParse(input, out var output);
+            output.ToString().Should().Be(input);
+        }
```
Comment (duncanp-sonar) [inline src/Core.UnitTests/SonarCompositeRuleIdTests.cs:60]: I forgot to check this test in before

File: src/Integration.Vsix.UnitTests/Analysis/DisableRuleCommandTests.cs [hunk 2]
```diff
@@ -38,6 +38,10 @@ namespace SonarLint.VisualStudio.Integration.Vsix.Analysis.UnitTests
     [TestClass]
     public class DisableRuleCommandTests
     {
+        private const int VisibleAndEnabled = (int)(OLECMDF.OLECMDF_SUPPORTED | OLECMDF.OLECMDF_ENABLED);
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/Analysis/DisableRuleCommandTests.cs:41]: I turned the expected values into constants to make the tests easier to read.

File: src/Integration.Vsix.UnitTests/Analysis/DisableRuleCommandTests.cs [hunk 3]
```diff
@@ -99,26 +106,31 @@ public void CheckStatusAndExecute_SingleCFamilyIssue_StandaloneMode_VisibleAndEn
...
         [TestMethod]
-        [DataRow(SonarLintMode.Connected)]
-        [DataRow(SonarLintMode.LegacyConnected)]
-        public void CheckStatus_SingleCFamilyIssue_ConnectedMode_VisibleButNotEnabled(SonarLintMode bindingMode)
+        [DataRow("c:S111", SonarLintMode.Connected, false)]
+        [DataRow("c:S111", SonarLintMode.LegacyConnected, false)]
+        [DataRow("cpp:S111", SonarLintMode.Connected, false)]
+        [DataRow("cpp:S111", SonarLintMode.LegacyConnected, false)]
+        [DataRow("javascript:S111", SonarLintMode.Connected, true)]
+        [DataRow("javascript:S111", SonarLintMode.LegacyConnected, true)]
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/Analysis/DisableRuleCommandTests.cs:125]: Special casing for JS - can still disable in connected modes for now.

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs [hunk 4]
```diff
@@ -95,10 +94,12 @@ private void QueryStatus(object sender, EventArgs args)
         {
             try
             {
-                // Show the menu item for Sonar errors...
-                bool isVisible = TryGetErrorCodeSync(errorList, out _);
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Analysis/DisableRuleCommand.cs:99]: Previously the behaviour was:
* we didn't supported disabling rules in all Sonar languages,
* but if disabling was supported it only worked in standalone mode.

Now the behaviour is slightly more complicated:
* we still don't support disabling rules in all Sonar languages;
* JS rules can always be disabled, regardless of mode,
* other supported languages (i.e. C, C++) can only be disabled in standalone mode.


`TryGetErrorCodeSync`

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs [hunk 5]
```diff
@@ -149,22 +150,18 @@ private void Execute(object sender, EventArgs e)
         /// 1) there is only one selected error in the error list and
         /// 2) it is a SonarLint error
         /// </summary>
-        internal /* for testing */ static  bool TryGetErrorCodeSync(IErrorList errorList, out string errorCode)
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Analysis/DisableRuleCommand.cs:152]: The old comment was inaccurate: there was a third condition: "3) the language supports disabling".
The comment is now accurate, as the enable/disable check is done elsewhere.

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs [hunk 6]
```diff
@@ -191,11 +188,28 @@ private static string FindErrorCodeForEntry(ITableEntryHandle handle)
             SonarRuleRepoKeys.Cpp
         };
 
-        private static bool IsDisablingRulesSupported(string errorCode)
+        private static void CalculateStatuses(SonarCompositeRuleId rule, SonarLintMode mode, out bool isVisible, out bool isEnabled)
         {
+            // Special case: JavaScript
+            // We don't support connected mode for JavaScript at the moment so we allow disabling
+            // in standalone mode. See #770.
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Analysis/DisableRuleCommand.cs:195]: Should be "... in connected mode."

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs (hunk 4 above)
Comment (rita-sonarsource) [reply src/Integration.Vsix/Analysis/DisableRuleCommand.cs:99]: Well, we could simplify the JS stuff, and combine it with C++. As in, if it's not standalone, return false.

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs [hunk 7]
```diff
@@ -191,11 +188,28 @@ private static string FindErrorCodeForEntry(ITableEntryHandle handle)
...
+            if (SonarRuleRepoKeys.AreEqual(SonarRuleRepoKeys.JavaScript, rule.RepoKey))
```
Comment (rita-sonarsource) [inline src/Integration.Vsix/Analysis/DisableRuleCommand.cs:196]: Is this block really needed? There's no connected mode in JS, so it should always work in the following block.

**Review by rita-sonarsource — APPROVED** [review]
Looks good, just the comment about JS's connected mode.

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs (hunk 4 above)
Comment (duncanp-sonar) [reply src/Integration.Vsix/Analysis/DisableRuleCommand.cs:99]: Yep, we could. That would mean you could disable JS rules in standalone mode, but you'd always have the default settings in connected mode i.e. if a user is already using connected mode and the JS rules are noisy, there's nothing they can do to stop it apart from stop using connected mode or uninstall SLVS.

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs (hunk 7 above)
Comment (duncanp-sonar) [reply src/Integration.Vsix/Analysis/DisableRuleCommand.cs:196]: Connected mode is set at solution level, so it is possible to have JS files in connected mode.

File: src/Integration.Vsix/Analysis/DisableRuleCommand.cs (hunk 4 above)
Comment (rita-sonarsource) [reply src/Integration.Vsix/Analysis/DisableRuleCommand.cs:99]: Ah I see, ok 👍

**Pull request merged**